### PR TITLE
autoscaler: add atleast 10% of storage with scaling

### PR DESCRIPTION
**Pull request by parth-gr**
during horizontal scaling after sometime,
we were just adding 3 more osd,
but in real use case we atleast need to scale the cluster by 10%, so atleast add 10% storage in the horizontal scaling

### CODE REVIEW

File: internal/controller/storageautoscaler/storageautoscaler_controller.go [hunk 1]
```diff
@@ -550,14 +552,56 @@ func calculateExpectedOsdSizeAndCount(storageCluster *ocsv1.StorageCluster, stor
 		// vertical scaling
 		// double the osd size
 		expectedOsdSize = *resource.NewQuantity(startOsdSize.Value()*2, startOsdSize.Format)
+
```
Comment (travisn) [inline internal/controller/storageautoscaler/storageautoscaler_controller.go:556]: This method has a lot of calculations. How about adding more unit tests for all the edge cases in the calculations to make sure we get the expected number of OSDs or detect limit is reached? The unit tests will give more confidence in the calculations than trying to code review it.

**Review by travisn — CHANGES_REQUESTED** [review]
How about updating the storage-auto-scale.md design doc?

**Comment by parth-gr** [comment]
Updated ready for final review

File: deploy/storageautoscaler.yaml [hunk 2]
```diff
@@ -8,6 +8,6 @@ spec:
     name: ocs-storagecluster
   deviceClass: ssd
   storageCapacityLimit: 100Ti
-  maxOsdSize: 8Ti
-  storageScalingThresholdPercent: 70
+  maxOsdSize: 100Gi
```
Comment (travisn) [inline deploy/storageautoscaler.yaml]: Revert this file from testing

File: internal/controller/storageautoscaler/storageautoscaler_controller.go [hunk 3]
```diff
@@ -550,14 +553,56 @@ func calculateExpectedOsdSizeAndCount(storageCluster *ocsv1.StorageCluster, stor
...
+			// calculate the maximum device sets we can add without exceeding the limit
+			remainingCapacity := storageAutoScaler.Spec.StorageCapacityLimit.Value() - startStorageCapacity.Value()
+			capacityPerDeviceSet := int64(replicaCount*deviceSetCount) * startOsdSize.Value()
+
+			maxDeviceSetsToAdd := int(remainingCapacity / capacityPerDeviceSet)
+
+			if maxDeviceSetsToAdd < 1 {
+				// cannot add even one device set, limit reached
+				limitReached = true
+				devicesToAdd = 1
```
Comment (travisn) [inline internal/controller/storageautoscaler/storageautoscaler_controller.go:593]: Will this cause an expansion even after reaching the limit in a previous expansion? It appears to always allowing expansion by 1 device set even at the max.

**Review by travisn — CHANGES_REQUESTED** [review]

File: internal/controller/storageautoscaler/storageautoscaler_controller.go (hunk 3 above)
Comment (parth-gr) [reply internal/controller/storageautoscaler/storageautoscaler_controller.go:593]: No, this is for the capcity limit alert message, where we need the next atleast available capacity to scale it.

```
			Message:   fmt.Sprintf("storage capacity limit reached for storageAutoScaler %q device class %q least expected storage capacity is %q", storageAutoScaler.Name, storageAutoScaler.Spec.DeviceClass, expectedStorageCapacity.String()),

```

**Review by travisn — APPROVED** [review]
/lgtm
